pkg/template/apis/template/types.go

```diff
@@ -76,6 +80,44 @@ type Parameter struct {
 	Required bool
 }
 
+// TemplateCompletion contains the requirements used to detect successful or
+// failed instantiations of a Template.
+type TemplateCompletion struct {
+	// deadlineSeconds is the number of seconds after which a template
+	// instantiation will be considered to be failed, if it has not already
+	// succeeded.
+	DeadlineSeconds *int64
+
+	// successRequirements hold all the requirements that must be met to
+	// consider the Template instantiation a success.  These requirements are
+	// ANDed together.
+	SuccessRequirements []TemplateCompletionRequirement
+
+	// failureRequirements hold the requirements that if any is met will cause
+	// the Template instantiation to be considered a failure.  These
+	// requirements are ORed together.
+	FailureRequirements []TemplateCompletionRequirement
+}
+
+// TemplateCompletionRequirement holds a single requirement that is matched to
+// partially determine success or failure of a Template instantiation.
+type TemplateCompletionRequirement struct {
+	// ref is a reference to an object created by the template.  Its name and
+	// namespace can contain references to parameters which will be substituted
+	// following the usual rules.
+	Ref kapi.ObjectReference
```

Comment:
Just seen that kapi.ObjectReference contains a member called FieldPath.  AFAICS it doesn't seem to be used much but I think it's intended to be a Kubernetes JSONPath expression - I don't think it'd be better, but it is a consideration to abandon the JSONPath/Condition/something else expansion point and just use that.